libraries/rush-lib/src/logic/operations/CacheableOperationPlugin.ts

```diff
@@ -0,0 +1,875 @@
+                });
+                if (cobuildCompletedState) {
+                  return cobuildCompletedState.status;
+                }
+                return status;
+              }
+            }
+          } else if (buildCacheContext.isCacheReadAllowed) {
+            buildCacheReadAttempted = !!projectBuildCache;
+            const restoreFromCacheSuccess: boolean | undefined =
+              await projectBuildCache?.tryRestoreFromCacheAsync(buildCacheTerminal);
+
+            if (restoreFromCacheSuccess) {
+              // Restore the original state of the operation without cache
+              await context._operationMetadataManager?.tryRestoreAsync({
+                terminal,
+                logPath,
+                errorLogPath
+              });
+              return OperationStatus.FromCache;
+            }
+          }
+          if (buildCacheContext.isSkipAllowed && !buildCacheReadAttempted) {
+            const isPackageUnchanged: boolean = !!(
+              lastProjectDeps &&
+              projectDeps &&
+              projectDeps.arguments === lastProjectDeps.arguments &&
+              _areShallowEqual(projectDeps.files, lastProjectDeps.files)
+            );
+
+            if (isPackageUnchanged) {
+              return OperationStatus.Skipped;
+            }
+          }
+
+          if (buildCacheContext.isCacheWriteAllowed && cobuildLock) {
+            const acquireSuccess: boolean = await cobuildLock.tryAcquireLockAsync();
+            if (acquireSuccess) {
+              // The operation may be used to marked remote executing, now change it to executing
+              context.status = OperationStatus.Executing;
+              runner.periodicCallback.addCallback(async () => {
+                await cobuildLock?.renewLockAsync();
+              });
+            } else {
+              // failed to acquire the lock, mark current operation to remote executing
+              context.stopwatch.reset();
+              return OperationStatus.RemoteExecuting;
+            }
+          }
+        })();
+        if (earlyReturnStatus) {
+          beforeExecuteContext.earlyReturnStatus = earlyReturnStatus;
+        }
+        return beforeExecuteContext;
+      }
+    );
+
+    runner.hooks.afterExecute.tapPromise(
+      PLUGIN_NAME,
+      async (afterExecuteContext: IOperationRunnerAfterExecuteContext) => {
+        const { context, status, taskIsSuccessful, logPath, errorLogPath } = afterExecuteContext;
+
+        const { cobuildLock, projectBuildCache, isCacheWriteAllowed, buildCacheTerminal } = buildCacheContext;
+
+        // Save the metadata to disk
+        const { duration: durationInSeconds } = context.stopwatch;
+        await context._operationMetadataManager?.saveAsync({
+          durationInSeconds,
+          cobuildContextId: cobuildLock?.cobuildConfiguration.cobuildContextId,
+          cobuildRunnerId: cobuildLock?.cobuildConfiguration.cobuildRunnerId,
+          logPath,
+          errorLogPath
```

Comment:
In order to not throw errors at runtime, these files must have been closed by the underlying runner when this code executes.

Reply:
Yeah. I believe the `projectWritable` for the shell process logging will be closed in the underlying runner right now.